mocks/consumer.go

```diff
@@ -289,7 +260,13 @@ func (pc *PartitionConsumer) HighWaterMarkOffset() int64 {
 // reasons forthis not to happen. ou can call ExpectMessagesDrainedOnClose so it will
 // verify that the channel is empty on close.
 func (pc *PartitionConsumer) YieldMessage(msg *sarama.ConsumerMessage) {
-	pc.expectations <- &consumerExpectation{Msg: msg}
+	atomic.AddInt64(&pc.highWaterMarkOffset, 1)
```

Comment:
I think it is necessary? `atomic.AddInt64` returns the new value, so you could collapse the atomic calls and reduce the window, but it would still be possible to interleave method calls such that messages ended up in the channel with the wrong order (e.g. offsets `1 2 3 4` instead of `1 2 3 4`). A lock seems the obvious way to force the `highWaterMarkOffset` change to occur synchronously with the channel write.

Note that the variable will still need to be edited atomically in order to avoid locking `HighWaterMarkOffset()` too.